src/main/java/com/aws/greengrass/secretmanager/SecretManager.java

```diff
@@ -69,74 +78,78 @@ public class SecretManager {
     private final AWSSecretClient secretClient;
     private final SecretStore<SecretDocument, AWSSecretResponse> secretStore;
     @Nullable
-    private Crypter crypter;
-    private final Result<SecretCryptoException> initialized = new Result<>();
+    private final AtomicReference<Crypter> crypter = new AtomicReference<>();
 
     /**
      * Constructor.
-     * @param secretClient client for aws secrets.
+     *
+     * @param secretClient    client for aws secrets.
      * @param securityService security service.
-     * @param dao          dao for persistent store.
+     * @param dao             dao for persistent store.
      */
     @Inject
     SecretManager(AWSSecretClient secretClient, SecurityService securityService, FileSecretStore dao,
-                  ExecutorService executor) {
+                  DeviceConfiguration deviceConfiguration) {
         this.secretStore = dao;
         this.secretClient = secretClient;
+        this.securityService = securityService;
+        deviceConfiguration.onAnyChange(((whatHappened, node) -> {
+            if (validUpdate(node, DEVICE_PARAM_CERTIFICATE_FILE_PATH) || validUpdate(node,
+                    DEVICE_PARAM_PRIVATE_KEY_PATH)) {
+                // TODO: If the device creds change, then cache and local store should be updated as well as the keys
+                //  used for encrypting and decrypting the secrets have changed.
+                synchronized (crypter) {
+                    crypter.set(null);
+                }
+            }
+        }));
+    }
+
+    private boolean validUpdate(Node node, String key) {
+        return node != null && node.childOf(key) && Utils.isNotEmpty(Coerce.toString(node)) && Files.exists(
+                Paths.get(key));
+    }
 
-        executor.execute(() -> loadCrypter(securityService));
+    private Crypter getCrypter() throws SecretCryptoException {
+        synchronized (Objects.requireNonNull(crypter)) {
+            if (crypter.get() == null) {
+                try {
+                    loadCrypter();
+                } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -118,3 +118,3 @@
                 } catch (Exception e) {
-                    throw new SecretCryptoException(&#34;Unable to load crypter&#34;);
+                    throw new SecretCryptoException(&#34;Unable to load crypter&#34;, e);
                 }

```